src/plugin/memory-provider.ts

```diff
@@ -53,133 +44,55 @@ export class MemoryProvider {
-    protected assertCapability(capability: keyof DebugProtocol.Capabilities, action: string): vscode.DebugSession {
-        const session = this.assertActiveSession(action);
-        if (!this.hasDebugCapability(session, capability)) {
-            throw new Error(`Cannot ${action}. Session does not have capability ${capability}.`);
-        }
-        return session;
-    }
-
-    protected hasDebugCapability(session: vscode.DebugSession, capability: keyof DebugProtocol.Capabilities): boolean {
-        return !!this.sessionDebugCapabilities.get(session.id)?.[capability];
-    }
-
-    protected hasClientCapability(session: vscode.DebugSession, capability: keyof DebugProtocol.InitializeRequestArguments): boolean {
-        return !!this.sessionClientCapabilities.get(session.id)?.[capability];
-    }
-
-    protected assertActiveSession(action: string): vscode.DebugSession {
-        if (!vscode.debug.activeDebugSession) {
-            throw new Error(`Cannot ${action}. No active debug session.`);
-        }
-        return vscode.debug.activeDebugSession;
+        context.subscriptions.push(vscode.debug.registerDebugAdapterTrackerFactory('*', { createDebugAdapterTracker }));
     }
 
     public async readMemory(args: DebugProtocol.ReadMemoryArguments): Promise<ReadMemoryResult> {
-        return sendRequest(this.assertCapability('supportsReadMemoryRequest', 'read memory'), 'readMemory', args);
+        return sendRequest(this.sessionTracker.assertDebugCapability(this.sessionTracker.activeSession, 'supportsReadMemoryRequest', 'read memory'), 'readMemory', args);
     }
 
     public async writeMemory(args: DebugProtocol.WriteMemoryArguments): Promise<WriteMemoryResult> {
-        const session = this.assertCapability('supportsWriteMemoryRequest', 'write memory');
+        const session = this.sessionTracker.assertDebugCapability(this.sessionTracker.activeSession, 'supportsWriteMemoryRequest', 'write memory');
         // Schedule a emit in case we don't retrieve a memory event
-        this.scheduledOnDidMemoryWriteEvents[args.memoryReference] = response => {
+        this.scheduledOnDidMemoryWriteEvents[session.id + '_' + args.memoryReference] = response => {
             // We only send out a custom event if we don't expect the client to handle the memory event
             // since our client is VS Code we can assume that they will always support this but better to be safe
             const offset = response?.offset ? (args.offset ?? 0) + response.offset : args.offset;
             const count = response?.bytesWritten ?? stringToBytesMemory(args.data).length;
-            this._onDidWriteMemory.fire({ memoryReference: args.memoryReference, offset, count });
+            // if our custom handler is active, let's fire the event ourselves
+            this.sessionTracker['fireSessionEvent'](session, 'memory-written', { memoryReference: args.memoryReference, offset, count });
```

Comment:
The bracket access here is a bit suspicious.

Reply:
I was not sure whether to make the method public out of habit but since this is a VS Code extension and not a framework, there is a lower risk of exposing "internal" methods. I made the method public since we actually use it from somewhere else, thanks!